File: evita/dom/forms/text_field_selection.cc
// Copyright (c) 2014 Project Vogue. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "evita/dom/forms/text_field_selection.h"

#include <algorithm>

#include "evita/dom/forms/text_field_control.h"

namespace dom {

//////////////////////////////////////////////////////////////////////
//
// TextFieldSelection
//
TextFieldSelection::TextFieldSelection(TextFieldControl* control)
    : ScriptableBase(),
      control_(control),
      anchor_offset_(0),
      focus_offset_(0) {}

TextFieldSelection::~TextFieldSelection() {}

void TextFieldSelection::set_anchor_offset(int anchor_offset) {
  auto const new_anchor_offset = NormalizeOffset(anchor_offset);
  if (anchor_offset_ == new_anchor_offset)
    return;
  anchor_offset_ = new_anchor_offset;
  control_->DidChangeSelection();
}

void TextFieldSelection::set_focus_offset(int focus_offset) {
  auto const new_focus_offset = NormalizeOffset(focus_offset);
  if (focus_offset_ == new_focus_offset)
    return;
  focus_offset_ = new_focus_offset;
  control_->DidChangeSelection();
}

void TextFieldSelection::DidChangeValue() {
  anchor_offset_ = NormalizeOffset(anchor_offset_);
  focus_offset_ = NormalizeOffset(focus_offset_);
}

int TextFieldSelection::NormalizeOffset(int offset) const {
  return std::max(0, std::min(offset, static_cast<int>(control_->length())));
}

}  // namespace dom

File: evita/dom/forms/text_field_selection.js
// Copyright (c) 2014 Project Vogue. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

Object.defineProperty(TextFieldSelection.prototype, 'end', {
  /**
   * @this {!TextFieldSelection}
   * @return {number}
   */
  get: function() {
    return Math.max(this.anchorOffset, this.focusOffset);
  }
});

Object.defineProperty(TextFieldSelection.prototype, 'start', {
  /**
   * @this {!TextFieldSelection}
   * @return {number}
   */
  get: function() {
    return Math.min(this.anchorOffset, this.focusOffset);
  }
});

Object.defineProperty(TextFieldSelection.prototype, 'text', {
  /**
   * @this {!TextFieldSelection}
   * @return {string}
   */
  get: function() {
    return this.collapsed ? '' :
        this.control.value.substring(this.start, this.end);
  },
  /**
   * @this {!TextFieldSelection}
   * @param {string} new_text
   * Note: We call setter |TextFieldControl.prototype.value_| to change value
   * of |TextFieldControl|. Because, setter |value| change selection to whole
   * value.
   */
  set: function(new_text) {
    var value = this.control.value;
    if (this.collapsed) {
      var offset = this.focusOffset;
      this.control.value_ = value.substr(0, offset) + new_text +
                            value.substr(offset);
      this.focusOffset = offset + new_text.length;
      return;
    }
    var start = this.start;
    var end = this.end;
    var end_is_focus = end == this.focusOffset;
    this.control.value_ = value.substr(0, start) + new_text + value.substr(end);
    // Relocate anchorOffset and focusOffset
    var new_end = new_text.length + start;
    if (end_is_focus)
      this.focusOffset = new_end;
    else
      this.anchorOffset = new_end;
  },
});

/** @param {number} offset */
global.TextFieldSelection.prototype.collapseTo = function(offset) {
  this.anchorOffset = this.focusOffset = offset;
};

/** @param {number} offset */
global.TextFieldSelection.prototype.extendTo = function(offset) {
  this.focusOffset = offset;
};
